indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java

```diff
@@ -214,61 +216,182 @@ public TaskStatus run(final TaskToolbox toolbox) throws Exception
+    long determineShardSpecsStartMillis = System.currentTimeMillis();
+
+    final Map<Interval, Optional<HyperLogLogCollector>> hllCollectors = collectIntervalsAndShardSpecs(
+        jsonMapper,
+        ingestionSchema,
+        firehoseFactory,
+        firehoseTempDir,
+        granularitySpec,
+        determineIntervals,
+        determineNumPartitions
+    );
+
+    final Map<Interval, List<ShardSpec>> intervalToShardSpecs = new HashMap<>();
+    final int defaultNumShards = tuningConfig.getNumShards() == null ? 1 : tuningConfig.getNumShards();
+    for (final Map.Entry<Interval, Optional<HyperLogLogCollector>> entry : hllCollectors.entrySet()) {
+      final Interval interval = entry.getKey();
+      final Optional<HyperLogLogCollector> collector = entry.getValue();
+
+      final int numShards;
+      if (determineNumPartitions) {
+        final long numRows = new Double(collector.get().estimateCardinality()).longValue();
+        numShards = (int) Math.ceil((double) numRows / tuningConfig.getTargetPartitionSize());
+        log.info("Estimated [%,d] rows of data for interval [%s], creating [%,d] shards", numRows, interval, numShards);
+      } else {
+        numShards = defaultNumShards;
+        log.info("Creating [%,d] shards for interval [%s]", numShards, interval);
       }
 
-      return shardSpecs;
+      final BiFunction<Integer, Integer, ShardSpec> shardSpecCreateFn = getShardSpecCreateFunction(
+          useExtendableShardSpec,
+          numShards,
+          jsonMapper
+      );
+
+      final List<ShardSpec> intervalShardSpecs = IntStream.range(0, numShards)
+                                                          .mapToObj(
+                                                              shardId -> shardSpecCreateFn.apply(shardId, numShards)
+                                                          ).collect(Collectors.toList());
+
+      intervalToShardSpecs.put(interval, intervalShardSpecs);
+    }
+    log.info("Found intervals and shardSpecs in %,dms", System.currentTimeMillis() - determineShardSpecsStartMillis);
+
+    if (useExtendableShardSpec) {
+      return createExtendableShardSpecs(intervalToShardSpecs);
+    } else {
+      return createNonExtendableShardSpecs(intervalToShardSpecs);
     }
+  }
 
-    // determine intervals containing data and prime HLL collectors
-    final Map<Interval, Optional<HyperLogLogCollector>> hllCollectors = Maps.newHashMap();
+  private static Map<Interval, Optional<HyperLogLogCollector>> collectIntervalsAndShardSpecs(
+      ObjectMapper jsonMapper,
+      IndexIngestionSpec ingestionSchema,
+      FirehoseFactory firehoseFactory,
+      File firehoseTempDir,
+      GranularitySpec granularitySpec,
+      boolean determineIntervals,
+      boolean determineNumPartitions
+  ) throws IOException
+  {
+    final Map<Interval, Optional<HyperLogLogCollector>> hllCollectors = new TreeMap<>(
+        Comparators.intervalsByStartThenEnd()
+    );
     int thrownAway = 0;
```

Comment:
This variable is never updated. The if below conditioning of this variable to be positive is dead code. See #7737.